recipes/libjpeg/all/conanfile.py

```diff
@@ -82,46 +82,64 @@ def generate(self):
-                lib = buildenv_vars.get("AR", "llvm-lib")
-                rc = compilers_from_conf.get("rc", buildenv_vars.get("RC", "llvm-rc"))
-                replace_in_file(self, "Win32.Mak", "cc     = cl", f"cc     = {cl}")
-                replace_in_file(self, "Win32.Mak", "link   = link", f"link   = {link}")
-                replace_in_file(self, "Win32.Mak", "implib = lib", f"implib = {lib}")
-                replace_in_file(self, "Win32.Mak", "rc     = Rc", f"rc     = {rc}")
-            # set flags directly in makefile.vc
-            # cflags are critical for the library. ldflags and ldlibs are only for binaries
-            if is_msvc_static_runtime(self):
-                replace_in_file(self, "makefile.vc", "(cvars)", "(cvarsmt)")
-                replace_in_file(self, "makefile.vc", "(conlibs)", "(conlibsmt)")
-            else:
-                replace_in_file(self, "makefile.vc", "(cvars)", "(cvarsdll)")
-                replace_in_file(self, "makefile.vc", "(conlibs)", "(conlibsdll)")
-            target = "{}/libjpeg.lib".format("shared" if self.options.shared else "static")
-            self.run("nmake -f makefile.vc {} {}".format(" ".join(make_args), target))
-
     def build(self):
         apply_conandata_patches(self)
-        if is_msvc(self) or self._is_clang_cl:
-            self._build_nmake()
+        if is_msvc(self):
+            with chdir(self, self.source_folder):
+                self.run("nmake /f makefile.vs setupcopy-v16")
+
+                # Rename target to 'libjpeg.lib' to match legacy behaviour (otherwise we break backwards compatibility)
+                jpeg_vcxproj = os.path.join(self.source_folder, "jpeg.vcxproj")
+                replace_in_file(self, jpeg_vcxproj, """<PropertyGroup Label="UserMacros" />""", 
+                                """ <PropertyGroup Label="UserMacros" /><PropertyGroup Label="TargetName"> <TargetName>libjpeg</TargetName></PropertyGroup>
+                                """)
+                
+                # Support static/shared
+                if self.options.shared:
+                    replace_in_file(self, jpeg_vcxproj,
+                        "<ConfigurationType>StaticLibrary</ConfigurationType>",
+                        "<ConfigurationType>DynamicLibrary</ConfigurationType>"
+                    )
+
+                # Inject conan-generated .props file
+                conantoolchain_props = os.path.join(self.generators_folder, MSBuildToolchain.filename)
+                replace_in_file(
+                    self, jpeg_vcxproj,
+                    """<Import Project="$(VCTargetsPath)\\Microsoft.Cpp.targets" />""",
+                    f"""<Import Project="{conantoolchain_props}" /><Import Project="$(VCTargetsPath)\\Microsoft.Cpp.targets" />""",
+                )
+
+                # Fix toolset 
+                toolset_before = "v142" if self.version in ["9d", "9e"] else "$(DefaultPlatformToolset)"
```

Comment:
In theory it is possible to override the ``vcxproj`` defined toolset with the ``conantoolchain.props`` one, as long as the inclusion is done in the right place. But not a problem with this either, should be good too.

Reply:
Good catch!
the right place is to import it right before `Microsoft.Cpp.props` - then the toolset version set by .conan props is respected, making this redundant :D - with the added benefit that the Visual Studio IDE then shows the Conan .props file correctly as well